### ISSUE TITLE
Inconsistencies in permission checking when creating a value for a property from another project

### ISSUE THREAD

**Issue by mrivoal** [dhlab-basel/Knora#408]
I have 2 projects : 

1. ***drawings-gods***
In this project, permissions are not defined yet, apart from the fact that there is this **user**: 
```
<http://data.knora.org/users/drawings-gods-test-user1> a knora-base:User ;
        knora-base:userid "ddd"^^xsd:string ;  
        knora-base:givenName "ddd"^^xsd:string ;  
        knora-base:familyName "ddd"^^xsd:string ;  
        knora-base:password "********"^^xsd:string ;
        knora-base:email "ddd@d.ch"^^xsd:string ;
        knora-base:preferredLanguage "fr"^^xsd:string ;
        knora-base:status "true"^^xsd:boolean ;
        knora-base:isInProject <http://data.knora.org/projects/drawings-gods> ;
        knora-base:isInProjectAdminGroup <http://data.knora.org/projects/drawings-gods> ;
        knora-base:isInGroup <http://data.knora.org/groups/drawings-gods-snf-team> ;
        knora-base:isInSystemAdminGroup "false"^^xsd:boolean .`
```

2. ***parole-religieuse***
In this project, there is a **user**: 

```
<http://data.knora.org/users/parole-religieuse-test-user1> a knora-base:User ;
        knora-base:userid "parole"^^xsd:string ;
        knora-base:givenName "parole"^^xsd:string ;
        knora-base:familyName "parole"^^xsd:string ;
        knora-base:password "**********"^^xsd:string ;
        knora-base:email "parole@unil.ch"^^xsd:string ;
        knora-base:preferredLanguage "fr"^^xsd:string ;
        knora-base:status "true"^^xsd:boolean ;
        knora-base:isInProject <http://data.knora.org/projects/parole-religieuse> ; ;
        knora-base:isInProjectAdminGroup <http://data.knora.org/projects/parole-religieuse> ;
        knora-base:isInSystemAdminGroup "false"^^xsd:boolean .
```
a **ProjectMember** group is also defined with this `default object access permission` : 
```
### Default Object Access Permissions on ProjectMember Group
<http://data.knora.org/permissions/parole-religieuse-o1> rdf:type knora-base:DefaultObjectAccessPermission ;
                      knora-base:forProject <http://data.knora.org/projects/parole-religieuse> ;
                      knora-base:forGroup knora-base:ProjectMember ;
                      knora-base:hasPermissions "CR knora-base:ProjectAdmin|D knora-base:ProjectMember|M knora-base:KnownUser|RV knora-base:UnknownUser"^^xsd:string .
```

I am currently running some tests to have a better understanding of what the Modify permission means, compared to Delete, etc., for values, for resources. So I came to this use case : 

User from **parole-religieuse** creates a resource `Affilliation` in parole-religieuse project, and a value for the property `affiliationDate`  which has `owl:minCardinality "1"`. User log out.

User from **drawings-gods** project searches for `Affiliation` resources and tries to add a value to `affiliationDate`. He actually can do it. But as soon as he did, he gets the message : 
```
error
Forbidden
{"status":3,"error":"org.knora.webapi.ForbiddenException: User http://data.knora.org/users/drawings-gods-test-user1 does not have permission to see value http://data.knora.org/zWr6JcMXSPOwIRXFoI_8yw/values/QHVkD_rPSHaRUf_yotzY-A","access":"NO_ACCESS"}
```
And he can't actually see the value he just created, it is as if it didn't exist where as, when logged as a user from parole-religieuse project, I am able to see the value he just created.

--------
Here are some questions : 

**question 1**: is it was is intended from Modify permission? Is a user allowed to create a new value for a property but NOT a new instance of a resource ? 

**question 2**: given the `default object access permission` above, a `knora-based:KnownUser` should View all the ressources and all the values in parole-religieuse project. However, after drawings-gods user created values in parole-religieuse project, he is not allowed to see them. 

**question 3** : how is it possible that the same user (drawings-gods user) is finally allowed to View only *some* values for the same property? At the end, when a parole-religieuse user was able to View 6 dates, the drawings-gods user could only View the values he did not created and that was only 2 of them.

**Comment by subotic** [comment]
> question 1: is it was is intended from Modify permission?

The behaviour depends on where the *modify* permission is attached. If it is attached on a **value**, then the user (or more precise the users inside the user group receiving the permission) are allowed to create new **versions** of an **existing** value. If the *modify* permission is residing on a resource, the those users are allowed to create new values (as opposed to a new version of an existing value).

In your example, the **parole-religieuse** user is creating a resource and a new value. So both the resource and the value should have ``M knora-base:KnownUser`` attached to them. This should be enough for the **drawing-gods** user to create a new value and/or a new version of an existing value. This is probably a bug. I will try to add a test tomorrow to replicate this issue.

> Is a user allowed to create a new value for a property but NOT a new instance of a resource ?

For a user to be allowed to create a new instance of a resource, he will need to have the administrative permission ``ProjectResourceCreateAllPermission``.

> question 2: given the default object access permission above, a knora-based:KnownUser should View all the ressources and all the values in parole-religieuse project. However, after drawings-gods user created values in parole-religieuse project, he is not allowed to see them.

This is an interesting case, for which I will also create a test. The default object access permissions are only applied when a resource or a value is created inside a project referenced by this permission. When the **drawing-gods** user creates a new value (not a new version), then this value is actually created inside the **drawing-gods** project, and as there are no default object access permissions defined for the **drawing-gods** project, then the newly created value will probably not have any permissions attached to it. We should probably have some sane default object access permissions applied in those cases, like ``CR knora-base:ProjectAdmin|M knora-base:ProjectMember``.

> question 3 : how is it possible that the same user (drawings-gods user) is finally allowed to View only some values for the same property? At the end, when a parole-religieuse user was able to View 6 dates, the drawings-gods user could only View the values he did not created and that was only 2 of them.

I think this has to do with the missing permissions from question 2.

Can you please also post the defined ``administrative permissions``?

I will create some tests tomorrow to replicate this setting in a new branch. We should know more then.

**Comment by subotic** [comment]
Work in progress: #409

**Comment by mrivoal** [comment]
Thanks for the clarification. I will need to read it more carefully but here are the administrative permissions you requested : 

```
<http://data.knora.org/permissions/parole-religieuse-a1> <http://www.knora.org/ontology/knora-base#forGroup> <http://www.knora.org/ontology/knora-base#ProjectAdmin> ;
	<http://www.knora.org/ontology/knora-base#forProject> <http://data.knora.org/projects/parole-religieuse> ;
	<http://www.knora.org/ontology/knora-base#hasPermissions> "ProjectResourceCreateAllPermission|ProjectAdminAllPermission" ;
	a <http://www.knora.org/ontology/knora-base#AdministrativePermission> .

<http://data.knora.org/permissions/parole-religieuse-a3> <http://www.knora.org/ontology/knora-base#forGroup> <http://www.knora.org/ontology/knora-base#ProjectMember> ;
	<http://www.knora.org/ontology/knora-base#forProject> <http://data.knora.org/projects/parole-religieuse> ;
	<http://www.knora.org/ontology/knora-base#hasPermissions> "ProjectResourceCreateAllPermission" ;
	a <http://www.knora.org/ontology/knora-base#AdministrativePermission> .
```

**Comment by mrivoal** [comment]
> question 2: given the default object access permission above, a knora-based:KnownUser should View all the ressources and all the values in parole-religieuse project. However, after drawings-gods user created values in parole-religieuse project, he is not allowed to see them.

> This is an interesting case, for which I will also create a test. The default object access permissions are only applied when a resource or a value is created inside a project referenced by this permission. When the **drawing-gods** user creates a new value (not a new version), then this value is actually created inside the **drawing-gods** project, and as there are no default object access permissions defined for the **drawing-gods** project, then the newly created value will probably not have any permissions attached to it. We should probably have some sane default object access permissions applied in those cases, like CR knora-base:ProjectAdmin|M knora-base:ProjectMember.

In fact, in my tests, when a **drawings-gods** user created a new value, the value was attached to the **parole-religieuse** project. But it might still be due to the fact that no administrative permission is set for **drawings-gods** project. I will set up these permissions tomorrow and let you know.

**Comment by subotic** [comment]
I have added a test that fails, meaning I can reproduce the issue.

What happens is, that the value created by the *drawings-gods* user has no permissions attached to it, and thus is not allowed to be seen by anybody.

Both ``knora-base:Resource`` and ``knora-base:Value`` have ``owl:maxCardinality "1"`` defined for the ``knora-base:hasPermissions`` property. What we didn't describe anywhere,  is what permission does one get if there are no permissions defined.

I will discuss this tomorrow with Ben, to see if this is a feature (correct behavior) or a bug ;-)

**Comment by subotic** [comment]
> In fact, in my tests, when a drawings-gods user created a new value, the value was attached to the parole-religieuse project.

Yes, this is also what I have done in the test.

Now I'm pretty sure that this is a bug. The value created by the **drawings-gods** user is attached to the **parole-religieuse** project and thus the default object access permissions of the **parole-religieuse** project should have been applied.

I'm still not sure how we should handle the case of the **drawings-gods** user creating a new value inside the **drawings-gods** project on a resource belonging to the **parole-religieuse** project. What permissions would the user need to have, to be able to do this, i.e. would he need to have **ProjectResourceCreateAll** for the **drawings-gods** project?

**Comment by benjamingeer** [comment]
> Both ``knora-base:Resource`` and ``knora-base:Value`` have ``owl:maxCardinality "1"`` defined for the ``knora-base:hasPermissions`` property. What we didn't describe anywhere,  is what permission does one get if there are no permissions defined.

I think Lukas and I talked about this at some point long ago, and he said he thought we should allow a resource or value to have no permissions at all, but I don't remember the reasoning for this. Maybe it's time to rethink it. I can't think of a case where it would make sense, and I've seen it happen several times unintentionally.

**Comment by mrivoal** [comment]
> I'm still not sure how we should handle the case of the **drawings-gods** user creating a new value inside the **drawings-gods** project on a resource belonging to the **parole-religieuse** project. What permissions would the user need to have, to be able to do this, i.e. would he need to have ProjectResourceCreateAll for the **drawings-gods** project?

It wasn't my intention. I was only trying to make sure that it was impossible for a user from another project (**drawings-gods**) to mess with the data belonging to another project (**parole-religieuse**), and to create and modify resources and values inside this project. More precisely, I wanted to investigate the Modify permission regarding resources and values in order to use it correctly.

**Comment by subotic** [comment]
> It wasn't my intention

I know. I'm a bit hijacking this issue with a situation that we didn't think of before, like this case :-)

**Comment by subotic** [comment]
> Now I'm pretty sure that this is a bug. The value created by the **drawings-gods** user is attached to the **parole-religieuse** project and thus the default object access permissions of the **parole-religieuse** project should have been applied.

I have found where the problem is. The current implementation looks at the groups the user is a member of for a certain project and then calculates the max default object access permissions. In our case, the **drawings-gods** user is not part of any group inside the **parole-religieuse** project, and thus results in an empty permissions string.

The solution would be to add the support for defining default object access permissions on the ``knora-base:KnownUser`` group inside the **parole-religieuse** project.

**Comment by mrivoal** [comment]
> The solution would be to add the support for defining default object access permissions on the knora-base:KnownUser group inside the **parole-religieuse** project.

But there is a default object permission on the  `knora-base:KnownUser `group inside the **parole-religieuse** project : 
```
### Default Object Access Permissions on ProjectMember Group
<http://data.knora.org/permissions/parole-religieuse-o1> rdf:type knora-base:DefaultObjectAccessPermission ;
                      knora-base:forProject <http://data.knora.org/projects/parole-religieuse> ;
                      knora-base:forGroup knora-base:ProjectMember ;
                      knora-base:hasPermissions "CR knora-base:ProjectAdmin|D knora-base:ProjectMember|M knora-base:KnownUser|RV knora-base:UnknownUser"^^xsd:string .
```
Do you mean that a default object permission should be defined inside the **drawings-gods** project ?

**Comment by subotic** [comment]
I mean something like this:

```
<http://data.knora.org/permissions/parole-religieuse-o2> rdf:type knora-base:DefaultObjectAccessPermission ;
        knora-base:forProject <http://data.knora.org/projects/parole-religieuse> ;
        knora-base:forGroup knora-base:KnownUser ;
        knora-base:hasPermissions "CR knora-base:ProjectAdmin|D knora-base:ProjectMember|M knora-base:KnownUser|RV knora-base:UnknownUser"^^xsd:string .
```
This would be additionally to the other default object access permission.

**Comment by mrivoal** [comment]
OK. I think I get it. 

Thanks.

**Comment by subotic** [comment]
should be fixed in #409
